central/auth/service/service_impl.go

```diff
@@ -89,3 +130,136 @@ func authStatusForID(id authn.Identity) (*v1.AuthStatus, error) {
+	storageConfigs, err := s.authDataStore.ListAuthM2MConfigs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.ListAuthMachineToMachineConfigResponse{Configs: storagetov1.AuthM2MConfigs(storageConfigs)}, nil
+}
+
+func (s *serviceImpl) GetAuthMachineToMachineConfig(ctx context.Context, id *v1.ResourceByID) (*v1.GetAuthMachineToMachineConfigResponse, error) {
+	config, exists, err := s.authDataStore.GetAuthM2MConfig(ctx, id.GetId())
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, errox.NotFound.Newf("auth machine to machine config with id %q", id.GetId())
+	}
+	return &v1.GetAuthMachineToMachineConfigResponse{Config: storagetov1.AuthM2MConfig(config)}, nil
+}
+
+func (s *serviceImpl) AddAuthMachineToMachineConfig(ctx context.Context, request *v1.AddAuthMachineToMachineConfigRequest) (*v1.AddAuthMachineToMachineConfigResponse, error) {
+	config := request.GetConfig()
+	resolveGitHubActionsIssuer(config)
+	if err := s.validateAuthMachineToMachineConfig(config, true); err != nil {
+		return nil, err
+	}
+	config.Id = uuid.NewV4().String()
+	storageConfig, err := s.authDataStore.AddAuthM2MConfig(ctx, v1tostorage.AuthM2MConfig(config))
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.AddAuthMachineToMachineConfigResponse{Config: storagetov1.AuthM2MConfig(storageConfig)}, nil
+}
+
+func (s *serviceImpl) UpdateAuthMachineToMachineConfig(ctx context.Context, request *v1.UpdateAuthMachineToMachineConfigRequest) (*v1.Empty, error) {
+	config := request.GetConfig()
+	resolveGitHubActionsIssuer(config)
+	if err := s.validateAuthMachineToMachineConfig(config, false); err != nil {
+		return nil, err
+	}
+
+	if err := s.authDataStore.UpdateAuthM2MConfig(ctx, v1tostorage.AuthM2MConfig(config)); err != nil {
+		return nil, err
+	}
+
+	return &v1.Empty{}, nil
+}
+
+func (s *serviceImpl) DeleteAuthMachineToMachineConfig(ctx context.Context, id *v1.ResourceByID) (*v1.Empty, error) {
+	if err := s.authDataStore.RemoveAuthM2MConfig(ctx, id.GetId()); err != nil {
+		return nil, errox.InvalidArgs.
+			Newf("could not delete auth machine to machine config with id %q", id.GetId()).CausedBy(err)
+	}
+	return &v1.Empty{}, nil
+}
+
+func (s *serviceImpl) ExchangeAuthMachineToMachineToken(_ context.Context, _ *v1.ExchangeAuthMachineToMachineTokenRequest) (*v1.ExchangeAuthMachineToMachineTokenResponse, error) {
+	return nil, errox.InvariantViolation.New("not yet implemented")
+}
+
+func (s *serviceImpl) validateAuthMachineToMachineConfig(config *v1.AuthMachineToMachineConfig, skipIDCheck bool) error {
+	if config == nil {
+		return errox.InvalidArgs.New("empty config given")
+	}
+	if config.GetId() == "" && !skipIDCheck {
+		return fmt.Errorf("%w: %w", errox.InvalidArgs, errEmptyID)
+	}
+
+	duration, err := time.ParseDuration(config.GetTokenExpirationDuration())
+	if err != nil {
+		return fmt.Errorf("%w: %w: %w", errox.InvalidArgs, errInvalidTokenExpiration, err)
```

Comment:
`errors.Wrapf`

Reply:
Does not work like this, the error chain is not preserved.